Add inginious-task-tester workflow template

nrybowski opened the pull request:
### Steps
- [x] Cache
  - [x] Pip packages
  - [x] APT packages
  - [x] INGInious and INGInious-containers repositories
  - [x] INGInious docker images
- [x] Installation and usage of additional task types from plugins
- [x] Remove TODOs when patches are merged in main INGInious repo
  - [x] Use main repo when patches are merged
  - [x] Remove default_install patch checkout
  - [x] Remove task_test patch checkout
  - [x] Remove latest task_test patch pull

### Todo in new feature requests
- Launch the workflow for different versions of INGInious and docker images
- Trigger the workflow when the version of inginious.info.ucl.ac.be / .org change (See https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads)

nrybowski committed:
Add inginious-task-tester workflow template

nrybowski committed:
[task-tester] Remove patch checkout

See https://github.com/UCL-INGI/INGInious/pull/860

nrybowski committed:
[task-tester] Revert temporary patches

See https://github.com/UCL-INGI/INGInious/pull/861

nrybowski commented:
@anthonygego This PR is ready for review.

anthonygego commented on workflow-templates/inginious-task-tester-container-builder.py:
```diff
@@ -0,0 +1,17 @@
+
+containers = os.environ.get('REQUIRED', 'base').split(' ')
+container_names = []
+for container in containers:
```
Here containers are iteratively built based on what the environment file contains. A choice has to made be made, either to:
- document and tell the users to specify the environments in order such that dependencies are met, otherwise it might behave strangely, crashing or pulling older versions on DockerHub.
- introspect dockerfiles recursively as it is done with `inginious-install` to solve dependencies automatically.

I'm OK with choosing the first option. The second can also be implemented later, as this is not blocking.

In the second case it would be nice to refactor the base code and allow to use it outside of `inginious-install`. This way, in case we add support for external Dockerfiles some day, those scripts would be ready too.

nrybowski replied:
As we discussed some time ago, the best solution would be building and pushing INGInious containers automatically on a registry (e.g. DockerHub and/or [GitHub](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry).). We could then simply pull the images from there rather than rebuilding them when the cache is not filled. This would also ensure uniformity between all the CI instances.

I will open a Feature Request on the main repo for that.

anthonygego replied:
Should I consider this as a choice for option 1 and that the PR is ready for merge then ?

nrybowski committed:
[inginious/task-tester] Add explicitly python3.11

The shipped version of python within the GitHub Workflow VM (Ubuntu20.04)
does not support typed syntax such as 'tule[int, list]'. Hence, the
explicit installation of a newwer python version is required.
